pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,628 @@
+				"The original datapath is still in-place, however it is recommended to retry the migration.")
+		return 0, -1, 0
+	}
+
+	v1Rules := filterRulesByPriority(rules, linux_defaults.RulePriorityEgress)
+	v2Rules := filterRulesByPriority(rules, linux_defaults.RulePriorityEgressv2)
+
+	// Keeps a record of the rules found which are seemingly partially
+	// migrated. User should be warned about these, so they can take action.
+	// These rules will be untouched.
+	var partial int
+
+	// (1) If compat=false and the current set of rules are under the older
+	// priority, then this is an upgrade migration.
+	//
+	// (2) If compat=false and the current set of rules are under the newer
+	// priority, then there is nothing to do.
+	//
+	// (3) If compat=true and the current set of rules are under the older
+	// priority, then there is nothing to do.
+	//
+	// (4) If compat=true and the current set of rules are under the newer
+	// priority, then this is a downgrade migration.
+
+	switch {
+	case compat && len(v1Rules) > 0 && len(v2Rules) > 0: // Unexpected older rules found.
+		partial = len(v1Rules)
+	case !compat && len(v1Rules) > 0 && len(v2Rules) > 0: // Unexpected newer rules found.
+		partial = len(v2Rules)
+	}
+
+	// Exit if there's nothing to do. Return the count of partially migrated
+	// rules.
+	switch {
+	case !compat && len(v1Rules) == 0 && len(v2Rules) > 0: // 2
+		fallthrough
+	case compat && len(v1Rules) > 0 && len(v2Rules) == 0: // 3
+		return 0, 0, partial
+	}
+
+	isUpgrade := !compat && len(v1Rules) > 0  // 1
+	isDowngrade := compat && len(v2Rules) > 0 // 4
+
+	// The following operation will be done on a per-rule basis (or
+	// per-endpoint, assuming that each egress rule has a unique IP addr
+	// associated with the endpoint).
+	//
+	// In both the upgrade and downgrade scenario, the following happens in a
+	// specific order to guarantee that any failure at any point won't cause
+	// connectivity disruption for the endpoint. Any errors encountered do not
+	// stop the migration process because we want to ensure that we conform to
+	// either the new state or the old state, and want to avoid being
+	// in-between datapath states.
+	//   1) Copy over new routes from the old routes
+	//   2) Insert new rule
+	//   3) Delete old rule
+	//   4) Delete old routes
+	// Doing (1) & (2) before (3) & (4) allows us to essentially perform an
+	// "atomic" swap-in for the new state.
+	//
+	// If a failure occurs at (1), then the old state can continue to service
+	// the endpoint. Similarly with (2) because routes without rules are likely
+	// to not have any effect.
+	//
+	// If a failure occurs, (3) or (4), then we already have the new state
+	// in-place to direct traffic for the endpoint. There is a slight nuance
+	// here; the older rule has a _higher_ priority which means it would be
+	// selected by Linux first, before reaching the new rule. Therefore, if we
+	// fail at (3), keep in mind we do have the new state in-place, but it goes
```

Comment:
**Warning**: it will get slightly confusing, but the TLDR is that the nuance doesn't really matter in the end.

I updated the comment to clarify. In both cases, it is possible to have both datapaths in-place and the rule priority will be the discriminator. 

To expand, if we fail at (3), then:

- Upgrade:
  - the original datapath is still in-place, using the old rule with a _higher_ priority. So no issues.
- Downgrade:
  - the downgrade was essentially completed and we'll be using the downgraded version because the "newer" rule has a _higher_ priority.

You could consider the downgrade case "more successful" if you go by which datapath is being used. In the upgrade case, the new datapath is not being used, but rather the original because it has a rule with a _higher_ priority.

Reply:
Isn't that a subtle difference though? On upgrade, failure at (3) means the datapath has not been upgraded, so you *should* clean up all the state we put in place during upgrade. On downgrade, failure at (3) means that the datapath has in fact been downgraded successfully to the older version, so then you *should not* remove the newly instantiated (downgraded) version of the datapath, otherwise you'll upgrade it again.

Reply:
Offline sync:

What Joe is pointing out is that we should have different logic for handling (3). Specifically for downgrade, we don't revert the work we've done up to this point; we can just call that it a success, because the endpoint is using the downgraded datapath, regardless of the leftover state.

In contrast, for upgrades, we should attempt to revert because the leftover state could block the user's next attempt at retying the upgrade.